### PR TITLE
template Management for dynamic prompt

### PR DESCRIPTION
**Pull request by htamahc**
### Proposed changes in this pull request

-     add a component to manage templates from the management console.
-     add an API to handle external requests.
-     UI component for the management console.

### When should this PR be merged

.

### Follow up actions

--

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceDataHolder.java [hunk 1]
```diff
@@ -88,6 +90,13 @@ private FrameworkServiceDataHolder() {
     public static FrameworkServiceDataHolder getInstance() {
         return instance;
     }
+    public void setTemplateManagerService(TemplateManager templateManagerService){
```
Comment (htamahc) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceDataHolder.java]: formatting issue

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TemplateMgtAPIClient.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import javax.net.ssl.HttpsURLConnection;
+import java.io.*;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Client for calling /api/identity/template/mgt/v1.0.0/templates/ with mutual ssl authentication
+ */
+public class TemplateMgtAPIClient {
+    private static final Log log = LogFactory.getLog(TemplateMgtAPIClient.class);
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TemplateMgtAPIClient.java]: Add a new line after class definition

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/pom.xml [hunk 3]
```diff
@@ -31,6 +32,17 @@
     <url>http://wso2.org</url>
 
     <dependencies>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-io</artifactId>
+            <version>1.3.2</version>
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/pom.xml]: Remove dependency versions from <dependencies>  and add to the dependencyManagement section in the root pom.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/pom.xml [hunk 4]
```diff
@@ -31,6 +32,17 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.framework</groupId>
+            <artifactId>org.wso2.carbon.identity.template.mgt</artifactId>
+            <version>5.12.199-SNAPSHOT</version>
+            <!--<scope>provided</scope>-->
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/pom.xml]: remove commented code

File: features/template-mgt/pom.xml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (madurangasiriwardena) [inline features/template-mgt/pom.xml]: Add license header to all the files.

File: features/template-mgt/pom.xml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    <name>WSO2 Carbon - Identity Template Management Feature Aggregator Module</name>
+    <url>http://wso2.org</url>
+    <modules>
+        <module>org.wso2.carbon.identity.template.mgt.feature</module>
+        <module>org.wso2.carbon.identity.template.mgt.server.feature</module>
+        <module>org.wso2.carbon.identity.template.mgt.ui.feature</module>
+    </modules>
+
+
+</project>
```
Comment (madurangasiriwardena) [inline features/template-mgt/pom.xml]: Add new line at the end of the file.

File: features/template-mgt/org.wso2.carbon.identity.template.mgt.ui.feature/pom.xml [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                            <propertiesFile>../../etc/feature.properties</propertiesFile>
+                            <adviceFile>
+                                <properties>
+                                    <propertyDef>org.wso2.carbon.p2.category.type:console</propertyDef>
+                                    <propertyDef>org.eclipse.equinox.p2.type.group:false</propertyDef>
+                                </properties>
+                            </adviceFile>
+                            <bundles>
+                                <bundleDef>org.wso2.carbon.identity.framework:org.wso2.carbon.identity.template.mgt.ui</bundleDef>
+                                <!--<bundleDef>org.wso2.carbon.identity.framework:org.wso2.carbon.identity.template.mgt</bundleDef>-->
```
Comment (madurangasiriwardena) [inline features/template-mgt/org.wso2.carbon.identity.template.mgt.ui.feature/pom.xml]: Remove commented code.

File: components/template-mgt/pom.xml [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+    </description>
+    <url>http://wso2.org</url>
+
+    <modules>
+        <module>org.wso2.carbon.identity.template.mgt</module>
+        <module>org.wso2.carbon.identity.template.mgt.endpoint</module>
+        <module>org.wso2.carbon.identity.template.mgt.ui</module>
+    </modules>
+
+    <!--<repositories>-->
```
Comment (madurangasiriwardena) [inline components/template-mgt/pom.xml]: remove commented code.

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateTest.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+public class TemplateTest {
+
+    private static final String templateName = "sampleTemplate";
+    private static final String description = "sample Description";
+    private static final String script = "sample Script";
+    private static Integer templateId = 1;
+    private static Integer tenantId = -1234;
+    Template template = new Template(templateId, tenantId, templateName, description, script);
+    Template template1 = new Template(templateName, description, script);
```
Comment (madurangasiriwardena) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateTest.java]: use a better variable name

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/internal/TemplateManagerComponent.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    protected void activate(ComponentContext componentContext) {
+
+        try {
+            BundleContext bundleContext = componentContext.getBundleContext();
+            DataSource dataSource = IdentityDatabaseUtil.getDataSource();
+            DatabaseUtil.getDBConnection(dataSource);
+            setDataSourceToDataHolder(dataSource);
+
+            bundleContext.registerService(TemplateManager.class, new TemplateManagerImpl(), null);
+//            bundleContext.registerService(TemplateManagerDAO.class, new TemplateManagerDAOImpl(), null);
```
Comment (madurangasiriwardena) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/internal/TemplateManagerComponent.java]: Remove commented code.

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/internal/TemplateManagerComponent.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            if (log.isDebugEnabled()) {
+                log.debug("Template Manager bundle is activated");
+            }
+        } catch (Throwable e) {
+            log.error("Error", e);
```
Comment (madurangasiriwardena) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/internal/TemplateManagerComponent.java]: Add better error message.

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/exception/TemplateManagementServerException.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * This class is used to define the server side errors which need to be handled.
+ */
+public class TemplateManagementServerException extends TemplateManagementException {
+
+    public TemplateManagementServerException(String message, String errorCode, Throwable cause) {
+
+        super(message, errorCode, cause);
+    }
+
```
Comment (madurangasiriwardena) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/exception/TemplateManagementServerException.java]: remove new line

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/exception/TemplateManagementException.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package org.wso2.carbon.identity.template.mgt.exception;
+
+/**
+ * Base exception for template management feature.
+ */
+public class TemplateManagementException extends Exception {
+
+    //    private static final long serialVersionUID = 2806215535431246551L;
```
Comment (madurangasiriwardena) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/exception/TemplateManagementException.java]: remove commented code.

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/dao/impl/TemplateManagerDAOImpl.java [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+        try {
+            jdbcTemplate.executeUpdate(TemplateMgtConstants.SqlQueries.INSERT_TEMPLATE, (preparedStatement -> {
+                preparedStatement.setInt(1, template.getTenantId());
+                preparedStatement.setString(2, template.getTemplateName());
+                preparedStatement.setString(3, template.getDescription());
+                try {
+                    InputStream inputStream = IOUtils.toInputStream(template.getTemplateScript());
+                    preparedStatement.setBinaryStream(4, inputStream, inputStream.available());
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (madurangasiriwardena) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/dao/impl/TemplateManagerDAOImpl.java]: don't use e.printStackTrace();. Use a logger and log the error with a proper error message.

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/dao/impl/TemplateManagerDAOImpl.java [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+     * @throws TemplateManagementException If error occurs while retrieving {@link Template}.
+     */
+    public Template getTemplateByName(String templateName, Integer tenantId) throws TemplateManagementException {
+
+        Template template;
+        JdbcTemplate jdbcTemplate = JdbcUtils.getNewTemplate();
+        try {
+            template = jdbcTemplate.fetchSingleRecord(GET_TEMPLATE_BY_NAME, ((resultSet, rowNumber) ->
+                    {
+                        Template template1 = null;
```
Comment (madurangasiriwardena) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/dao/impl/TemplateManagerDAOImpl.java]: use a proper variable name.

**Comment by madurangasiriwardena** [comment]
@htamahc thanks you for your contribution. 

Please fix the comments added above. And fix below as well.
- Remove the files added for the demo.
- Remove the files (js, css etc) added for the code mirror and use the existing files.

After fixing the issues mentioned please re-open the PR.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/template-mapper.jsp [hunk 16]
```diff
@@ -22,4 +22,8 @@
     Map<String, String> templateMap = new HashMap<>();
     templateMap.put("genericForm", "templates/genericForm.jsp");
     templateMap.put("username", "templates/username.jsp");
+    templateMap.put("verify_Age", "templates/verify_Age.jsp");
+    templateMap.put("verifyAge", "templates/verifyAge.jsp");
```
Comment (htamahc) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/template-mapper.jsp]: remove

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/pom.xml [hunk 17]
```diff
@@ -32,6 +32,11 @@
     <name>WSO2 Carbon - Identity Application Authentication Framework</name>
 
     <dependencies>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.framework</groupId>
+            <artifactId>org.wso2.carbon.identity.template.mgt</artifactId>
+            <version>5.12.199-SNAPSHOT</version>
```
Comment (htamahc) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/pom.xml]: Update version

File: components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/gen/java/org/wso2/carbon/identity/template/mgt/endpoint/ApiResponseMessage.java [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    public String getMessage() {
+
+        return message;
+    }
+
+    public void setMessage(String message) {
+
+        this.message = message;
+    }
+}
```
Comment (htamahc) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/gen/java/org/wso2/carbon/identity/template/mgt/endpoint/ApiResponseMessage.java]: add new line

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/codemirror/addon/hint/htmlhint.js [hunk 19]
```diff
@@ -0,0 +1,482 @@
...
+            if (s && "title" === o) {
+                var l = i.lastEvent;
+                ("text" !== l.type || "text" === l.type && /^\s*$/.test(l.raw) === !0) && t.error("<title></title> must not be empty.", i.line, i.col, r, i.raw)
+            } else "head" === o && (s === !1 && t.error("<title> must be present in <head> tag.", i.line, i.col, r, i.raw), e.removeListener("tagstart", a), e.removeListener("tagend", n))
+        }
+
+        var r = this, i = !1, s = !1;
+        e.addListener("tagstart", a), e.addListener("tagend", n)
+    }
+});
```
Comment (htamahc) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/codemirror/addon/hint/htmlhint.js]: Add a new line at the end of each file

**Pull request closed without merging**